### PR TITLE
fix(rest): License information generation based on attachment usages from rest

### PR DESCRIPTION
**Pull request by smrutis1**
* License information generation from rest endpoint based on the persisted attachment usages.

Closes: #686

Signed-off-by: Smruti Sahoo <smruti.sahoo@siemens.com>

### CODE REVIEW

**Comment by mcjaeger** [comment]
resterted build, although it seems to be OK:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 08:51 min
[INFO] Finished at: 2019-10-01T16:32:45Z
[INFO] Final Memory: 183M/776M
[INFO] ------------------------------------------------------------------------
*** Error in `/usr/lib/jvm/java-8-openjdk-amd64/bin/java': double free or corruption (fasttop): 0x00007f65de0c1270 ***
/home/travis/.travis/functions: line 109:  9093 Aborted                 (core dumped) mvn --fail-at-end $MVN_ARGS
The command "mvn --fail-at-end $MVN_ARGS" exited with 134.
cache.2
store build cache
Done. Your build exited with 1.
/home/travis/.travis/functions: line 508: wait: pid 2658 is not a child of this shell
```

**Comment by nutanv1contr** [comment]
Tested the PR. Below is my observation.

1. Open project>> goto atachment usage and select an attachment then generate license info from UI and rest end point. both are consistent.
2. Go to license info generation page exclude few of the license and save the changes. license info generation from UI is generating license document without the excluded licenses. but from rest end point it is considering all the licenses from the selected attachment.

**Comment by smrutis1** [comment]
@nutanv1contr , Thanks for pointing it out. Fixed the reported issue.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/Sw360ComponentService.java [hunk 1]
```diff
@@ -51,6 +52,15 @@
         return sw360ComponentClient.getComponentSummary(sw360User);
     }
 
+    public Release getReleaseById(String id, User sw360User) {
+        try {
+            ComponentService.Iface sw360ComponentClient = getThriftComponentClient();
+            return sw360ComponentClient.getReleaseById(id, sw360User);
+        }catch (TException e) {
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/component/Sw360ComponentService.java]: Done.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java [hunk 2]
```diff
@@ -273,22 +284,42 @@ public void downloadLicenseInfo(@PathVariable("id") String id,
...
+                            excludedLicensesPerAttachments.put(attachment.getAttachmentContentId(), licenseNameWithText);
+                            selectedReleaseAndAttachmentIds.get(releaseLink.getId()).add(attachment.getAttachmentContentId());
+                    }
                 }
-            }
-        }
-
-        final Map<String, Set<LicenseNameWithText>> excludedLicenses = new HashMap<>(); // TODO: implement method to determine excluded licenses
+            });
+        }));
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java]: Done.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java [hunk 3]
```diff
@@ -150,4 +168,58 @@ public Project convertToEmbeddedWithExternalIds(Project sw360Object) {
...
+                    .map(releaseLinkMapper)
+                    .collect(Collectors.toList());
+            projectLink.setLinkedReleases(mappedReleaseLinks);
+            return projectLink;
+        };
+    }
+
+    protected List<ProjectLink> createLinkedProjects(Project project, Function<ProjectLink, ProjectLink> projectLinkMapper, boolean deep,
+            User user) {
+        final Collection<ProjectLink> linkedProjects = flattenProjectLinkTree(getLinkedProjects(project, deep, user));
```
Comment (akapti) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java]: Can we re-use `flattenProjectLinkTree` and `getLinkedProjects` method present in `SW360Utils`?

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java [hunk 4]
```diff
@@ -33,12 +34,16 @@
...
-import org.springframework.security.oauth2.provider.OAuth2Authentication;
 import org.springframework.stereotype.Service;
 import org.springframework.util.FileCopyUtils;
 import org.springframework.web.multipart.MultipartFile;
 
+import com.google.common.collect.Sets;
+
 import javax.servlet.http.HttpServletResponse;
+
+import static org.eclipse.sw360.datahandler.common.WrappedException.wrapTException;
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java]: Done.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java (hunk 3 above)
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/Sw360ProjectService.java]: Done

**Pull request merged**